rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2778 @@
+The *only* item defined by this RFC requiring special compiler support is `TransmuteFrom`. To realize this RFC's proposal of safe transmutation between different types, this item will require compiler support. However, the *most* minimal acceptable implementation of `TransmuteFrom` can be achieved entirely in-language:
+
+```rust
+/// A type is transmutable into itself.
+unsafe impl<T, Neglect> TransmuteFrom<T, Neglect> for T
+where
+    Neglect: UnsafeTransmuteOptions
+{}
+
+/// A transmutation is *stable* if...
+unsafe impl<Src, Dst> TransmuteFrom<Src> for Dst
+where
+    Src: PromiseTransmutableInto,
+    Dst: PromiseTransmutableFrom,
+    <Dst as PromiseTransmutableFrom>::Archetype:
+        TransmuteFrom<
+            <Src as PromiseTransmutableInto>::Archetype,
+            NeglectStability
+        >
+{}
+```
+This [minimal implementation][minimal-impl] is sufficient for convincing the compiler to accept basic stability declarations, such as those of Rust's primitive types. It is *insufficient* for making the compiler accept transmutations between *different* types (and, consequently, complex stability declarations). Implementers should use this as a starting point.
+
+### Listing for Initial, Minimal Implementation
+[minimal-impl]: #Listing-for-Initial-Minimal-Implementation
+This listing is both a minimal-viable implementation of this RFC (excepting the automatic derives) and the **canonical specification** of this RFC's API surface ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=dee36fb5231c9e00b03b424e204cd1a9)):
+```rust
+#![feature(const_generics)] // for stability declarations on `[T; N]`
+#![feature(decl_macro)] // for stub implementations of derives
+#![feature(never_type)] // for stability declarations on `!`
+#![allow(unused_unsafe, incomplete_features)]
+
+/// Transmutation conversions.
+// suggested location: `core::convert`
+pub mod transmute {
+
+    use {options::*, stability::*};
+
+    /// Reinterprets the bits of a value of one type as another type, safely.
+    #[inline(always)]
+    pub /*const*/ fn safe_transmute<Src, Dst, Neglect>(src: Src) -> Dst
+    where
+        Src: TransmuteInto<Dst, Neglect>,
+        Neglect: SafeTransmuteOptions
+    {
+        src.transmute_into()
+    }
+
+    /// Reinterprets the bits of a value of one type as another type, potentially unsafely.
+    #[inline(always)]
+    pub /*const*/ unsafe fn unsafe_transmute<Src, Dst, Neglect>(src: Src) -> Dst
+    where
+        Src: TransmuteInto<Dst, Neglect>,
+        Neglect: UnsafeTransmuteOptions
+    {
+        unsafe { src.unsafe_transmute_into() }
+    }
+
+    /// `Self: TransmuteInto<Dst, Neglect`, if the compiler accepts the stability,
+    /// safety, and soundness of transmuting `Self` into `Dst`, notwithstanding
+    /// a given set of static checks to `Neglect`.
+    pub unsafe trait TransmuteInto<Dst: ?Sized, Neglect = ()>
+    where
+        Neglect: UnsafeTransmuteOptions,
```

Comment:
**Unresolved question: *Should this additional bound be added here:***
```suggestion
    where
        Dst: TransmuteFrom<Self, Neglect>,
        Neglect: UnsafeTransmuteOptions,
```
A bound that communicates that `TransmuteInto` is the reciprocal of `TransmuteFrom` seems desirable, but introducing it [requires a lot of seemingly redundant bounds elsewhere](https://gist.github.com/jswrenn/19aad724cb2e804fae6abd31a12988ff/revisions#diff-1fb9bba94efc1d84c71b0bd852912496). The [`Into`](https://doc.rust-lang.org/std/convert/trait.Into.html) trait doesn't have this reflexivity bound, so maybe it's not actually a desirable thing to have?